### PR TITLE
Remove code duplication in ActiveSupport::Cache

### PR DESCRIPTION
**Pull request by bogdan**
### Summary

Remove code duplication in the implementation of the cache key

### CODE REVIEW

File: actionpack/lib/abstract_controller/caching.rb [hunk 1]
```diff
@@ -57,7 +57,7 @@ def view_cache_dependencies
       # Convenience accessor.
       def cache(key, options = {}, &block) # :doc:
         if cache_configured?
-          cache_store.fetch(ActiveSupport::Cache.expand_cache_key(key, :controller), options, &block)
+          cache_store.fetch([:controller, key], options, &block)
```
Comment (matthewd) [inline actionpack/lib/abstract_controller/caching.rb]: This sounds like a behavioural change

File: activesupport/lib/active_support/cache.rb [hunk 2]
```diff
@@ -77,24 +77,22 @@ def lookup_store(*store_option)
...
       def expand_cache_key(key, namespace = nil)
-        expanded_cache_key = (namespace ? "#{namespace}/" : "").dup
-
-        if prefix = ENV["RAILS_CACHE_ID"] || ENV["RAILS_APP_VERSION"]
-          expanded_cache_key << "#{prefix}/"
-        end
-
-        expanded_cache_key << retrieve_cache_key(key)
-        expanded_cache_key
+        [namespace, ENV["RAILS_CACHE_ID"] || ENV["RAILS_APP_VERSION"], retrieve_cache_key(key)].compact.join("/")
```
Comment (matthewd) [inline activesupport/lib/active_support/cache.rb]: This is more allocations in a reasonably hot path, and also seems less readable to me

File: activesupport/lib/active_support/cache.rb [hunk 3]
```diff
@@ -565,33 +563,13 @@ def merged_options(call_options)
...
           prefix = namespace.is_a?(Proc) ? namespace.call : namespace
           key = "#{prefix}:#{key}" if prefix
           key
         end
 
-        # Expands key to be a consistent string value. Invokes +cache_key+ if
-        # object responds to +cache_key+. Otherwise, +to_param+ method will be
-        # called. If the key is a Hash, then keys will be sorted alphabetically.
-        def expanded_key(key)
-          return key.cache_key.to_s if key.respond_to?(:cache_key)
```
Comment (matthewd) [inline activesupport/lib/active_support/cache.rb]: Is there relevant history to why this method wasn't taught about `cache_key_with_version`?

File: actionpack/lib/abstract_controller/caching.rb (hunk 1 above)
Comment (bogdan) [reply actionpack/lib/abstract_controller/caching.rb]: Nope: in previous implementation `controller` as namespace prepends the key just in the same way as if the key has two parts in current implementation.

File: activesupport/lib/active_support/cache.rb (hunk 2 above)
Comment (bogdan) [reply activesupport/lib/active_support/cache.rb]: This is less allocation: in previous impelementation you allocate many strings. In this implementation you allocate one array and the dup of it (via `compact`).
We can change `compact` to `tap(&:compact!)` if you care about allocations.

Put 👎  and I will revert to previous implementation. As I am not super concerned.

I am not sure what is "a reasonably hot path". Can you explain?

File: activesupport/lib/active_support/cache.rb (hunk 3 above)
Comment (bogdan) [reply activesupport/lib/active_support/cache.rb]: It goes back up to 7 years: there wasn't much change there and rails was not driven by united mind in that era.

File: actionpack/lib/abstract_controller/caching.rb (hunk 1 above)
Comment (matthewd) [reply actionpack/lib/abstract_controller/caching.rb]: What about the ENV value?

File: activesupport/lib/active_support/cache.rb (hunk 2 above)
Comment (matthewd) [reply activesupport/lib/active_support/cache.rb]: Previously it was one string in `expanded_cache_key` (two if `namespace`), plus one more if `prefix`. With this change it's the array literal, the compacted array, and the joined string.

"hot path" = code-path that's called very frequently, deserving of performance concern; reasonably (in this context) = not very, but somewhat [hot].

**Pull request merged**